proposals/0723-qualified-strings.rst

```diff
@@ -0,0 +1,289 @@
+
+* Type inference ambiguity.
+
+  * Consider the following code:
+
+    ::
+
+      output :: IsString s => s -> IO ()
+
+      main = do
+        output "hello"
+        output "world"
+
+        -- output $ T.replace " " "_" input
+
+    This originally works with no extensions, due to the string literals being typed to concrete ``String``. But say the developer wants to add a call to ``T.replace`` and use ``Text`` literals; adding ``OverloadedStrings`` would cause ambiguity to the existing locations because they are now no longer concretely ``String``.
+
+This proposal would allow using a module qualifier to say precisely which function to desugar to, rather than using type classes, in a similar manner as ``-XQualifiedDo``. This would allow writing the previous code as
+
+::
+
+  {-# LANGUAGE QualifiedStrings #-}
+
+  main = do
+    output "hello"
+    output "world"
+
+    output $ T.replace T." " T."_" input
+
+The existing locations would continue working as ``String``, while the new line would unambiguously desugar to ``T.replace (T.pack " ") (T.pack "_") input``.
+
+Proposed Change Specification
+-----------------------------
+
+Introduce ``-XQualifiedStrings`` that desugars literal string syntax to function calls in a similar way to ``-XQualifiedDo`` (`docs <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/qualified_do.html>`_, `proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0216-qualified-do.rst>`_).
+
+As long as the desugared expressions/patterns type check, users are free to define these functions however they want. No whitespace is allowed between the ``.`` and the module name / literal.
+
+Currently, string literals have the following desugaring:
+
+.. list-table::
+    :align: left
+
+    * - **Expression**
+      - **Enabled extensions**
+      - **Desugared expression syntax**
+    * - ``"hello"``
+      -
+      - ``"hello"``
+    * - ``"hello"``
+      - ``-XOverloadedStrings``
+      - ``GHC.Exts.fromString "hello"``
+    * - ``"""hello"""``
+      - ``-XMultilineStrings``
+      - ``"hello"``
+
+With ``-XQualifiedStrings``, we gain the following syntaxes:
+
+.. list-table::
+    :align: left
+
+    * - **New expression syntax**
+      - **Additional extensions**
+      - **Desugared expression syntax**
+    * - ``M."asdf"``
+      -
+      - ``M.fromString "asdf"``
+    * - ``M."""asdf"""``
+      - ``-XMultilineStrings``
+      - ``M.fromString "asdf"``
+
+.. list-table::
+    :align: left
+
+    * - **New pattern syntax**
+      - **Additional extensions**
+      - **Desugared pattern syntax**
+    * - ``M."asdf"``
+      -
+      - ``((== M.fromString "asdf") -> True)``
+    * - ``M."""asdf"""``
+      - ``-XMultilineStrings``
+      - ``((== M.fromString "asdf") -> True)``
+
+It is highly recommended that all types with ``IsString`` instances defined provide a module with the below definitions, to enable locally-scoped overloading over ``-XOverloadedStrings``:
```

Comment:
Still confusing.  Maybe the code can be
```
module M where

data MyString = ...

instance S.IsString MyString where   -- Makes MyString useful with OverloadedStrings
  fromString = ... 

fromString :: String -> MyString     -- Ensures that QualifiedStrings has same meaning as OverloadedStrings
fromString = S.fromString           -- for MyString
```

Reply:
hm I was assuming people would want to put `fromString` in a separate module from their main types. But sure, I could do that